### PR TITLE
Adding NewBookmarksDaoTest

### LINKED ISSUE #2012: Add unit tests for NewBookmarksDao
**Type of Test**

* [x] Unit Test
* [ ] UI Test

**Particular area and files to be tested**

NewBookmarksDao.kt

**Current Status of the tests**

Coverage is at 33.13%

**Particular Behaviour to be tested**

Functions: getBookmarks, getCurrentZimBookmarksUrl, bookmarkUrlsForCurrentBook, saveBookmark, deleteBookmarks, deleteBookmarks, migrationInsert

### PR DESCRIPTION
**Pull request by akkywadhwa**
Fixes #2012

**Screenshots** 
N/A

### CODE REVIEW

**Comment by akkywadhwa** [comment]
@macgills I am facing issues in mocking : - 
 every {
       queryBuilder.equal(
         BookmarkEntity_.zimId,
         zimFileReader?.id ?: ""
  )
} returns queryBuilder
What do I mock zimFileReader here?

**Comment by akkywadhwa** [comment]
> @macgills I am facing issues in mocking : -
> every {
> queryBuilder.equal(
> BookmarkEntity_.zimId,
> zimFileReader?.id ?: ""
> )
> } returns queryBuilder
> What do I mock zimFileReader here?

I found mocked zimFileReader in StorageObserverTest.kt
Here is the code:
`
mockkStatic(CoreApp::class)
every { CoreApp.getInstance().packageName } returns "pkg"
zimFileReader = mockk()
`
But according to me testing can be done better by taking a local zim file. 
Please let me know how should I proceed?

**Comment by macgills** [comment]
@akkywadhwa you need to test that the query gets built with the id ZimFileReader gives you, for this scenario use a mock.
You also need to test the scenario that a null ZimFileReader returns `""`

**Comment by akkywadhwa** [comment]
> @akkywadhwa you need to test that the query gets built with the id ZimFileReader gives you, for this scenario use a mock.
> You also need to test the scenario that a null ZimFileReader returns `""`

Okay. I wanted to know if I should use a dummy zim file for the test?

**Comment by macgills** [comment]
@akkywadhwa try `gradlew jacocoTestDebugUnitTestReport` otherwise you can rely on the reports that codecov posts here

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
+    val query: Query<BookmarkEntity> = mockk(relaxed = true)
+    every { queryBuilder.build() } returns query
+    newBookmarksDao.getBookmarks(false, zimFileReader)
+    verify { queryBuilder.build() }
+  }
+
+  @Test
+  fun `get bookmarks from current book`() {
+    mockkStatic(CoreApp::class)
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:60]: why?

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  fun `get bookmarks without current book`() {
+    mockkStatic(CoreApp::class)
+    every { CoreApp.getInstance().packageName } returns "pkg"
+    val zimFileReader: ZimFileReader = mockk()
+    val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
+    every { box.query() } returns queryBuilder
+    every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
+    val query: Query<BookmarkEntity> = mockk(relaxed = true)
+    every { queryBuilder.build() } returns query
+    newBookmarksDao.getBookmarks(false, zimFileReader)
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:54]: there is no assertion happening, 

```
every { query.find() } returns bookmarkEntities
```
where `bookmarkEntities` is a list of maybe even just 1 item

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    verify { queryBuilder.build() }
+  }
+
+  @Test
+  fun `get bookmarks from current book`() {
+    mockkStatic(CoreApp::class)
+    every { CoreApp.getInstance().packageName } returns "pkg"
+    val zimFileReader: ZimFileReader = mockk()
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:62]: there is no test when `fromCurrentBook` is true and ZimFileReader is null

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
+    val query: Query<BookmarkEntity> = mockk(relaxed = true)
+    every { queryBuilder.build() } returns query
+    newBookmarksDao.getBookmarks(true, zimFileReader)
+    verify { queryBuilder.equal(BookmarkEntity_.zimName, zimFileReader.name) }
+    verify { queryBuilder.build() }
+  }
+
+  @Test
+  fun `get current zim bookmarks url`() {
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:81]: this doesn't call any dao methods

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        zimFileReader.id
+      )
+    } returns queryBuilder
+    every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
+    val query: Query<BookmarkEntity> = mockk(relaxed = true)
+    every { queryBuilder.build() } returns query
+    val queryResult: List<BookmarkEntity> = mockk()
+    mockkStatic(RxQuery::class)
+    every { RxQuery.observable(query) } returns Observable.just(queryResult)
+    newBookmarksDao.bookmarkUrlsForCurrentBook(zimFileReader)
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:133]: for things that return flowables you call `.test()` on them and then you get access to the assert function on a flowable.

**Comment by akkywadhwa** [comment]
> @akkywadhwa try `gradlew jacocoTestDebugUnitTestReport` otherwise you can rely on the reports that codecov posts here

@macgills Coverage report is failing here.

**Comment by macgills** [comment]
@akkywadhwa failing with what?

image


runs fine for me on develop

you would want to look under core>build>jacoco>index.html

The console spout a bunch of errors but it actually does everything just fine

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt (hunk 2 above)
Comment (macgills) [reply core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:49]: >there is no assertion happening

I want you to be verifying the result of `newBookmarksDao.getBookmarks(false, zimFileReader)` was my point

**Comment by akkywadhwa** [comment]
> @akkywadhwa failing with what?
> 
> image
> 
> runs fine for me on develop
> 
> you would want to look under core>build>jacoco>index.html
> 
> The console spout a bunch of errors but it actually does everything just fine

See below. "Coverage Reporting / coverageReport (21) (pull_request)" is failing.

**Comment by macgills** [comment]
I was lazy in my example tests and didn't use chained mocks properly so I gave up on them due to limited time but I found myself on the documentation and we really should use them as they'd make these tests a lot prettier.

The trick with [chained calls](https://mockk.io/#chained-calls) is that for generic return types we have to use `hint`

**Comment by macgills** [comment]
> Please see below. "Coverage Reporting / coverageReport (21) (pull_request)" is failing.

You generally don't have to care as long as 1 of them passes, instrumented tests are right bastards most of the time. Worry if both fail

File: core/src/sharedTestFunctions/java/org/kiwix/sharedFunctions/TestModelFunctions.kt [hunk 6]
```diff
@@ -164,3 +165,14 @@ fun libraryNetworkEntity(books: List<Book> = emptyList()) = LibraryNetworkEntity
 
 fun recentSearchEntity(id: Long = 0L, searchTerm: String = "", zimId: String = "") =
   RecentSearchEntity(id, searchTerm, zimId)
+
+fun bookmarkItem(
+  databaseId: Long = 0L,
+  zimId: String = "",
+  zimName: String = "",
+  zimFilePath: String?,
```
Comment (macgills) [inline core/src/sharedTestFunctions/java/org/kiwix/sharedFunctions/TestModelFunctions.kt:173]: to keep with the style of the rest of these functions you should supply default parameters for all fields of the object

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+package org.kiwix.kiwixmobile.core.dao
+
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import io.objectbox.Box
+import io.objectbox.query.Query
+import io.objectbox.query.QueryBuilder
+import org.junit.Test
```
[the remark this answers is not in the document]
Comment (akkywadhwa) [reply core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:27]: No issues! I will make the changes. 👍

**Comment by macgills** [comment]
image

Sweet sweet coverage.

We aren't at 100 yet though and that is what I would like to see.

Also I think you may have overlooked a few of my comments (github can get very messy)

>I was lazy in my example tests and didn't use chained mocks properly so I gave up on them due to limited time but I found myself on the documentation and we really should use them as they'd make these tests a lot prettier.
The trick with chained calls is that for generic return types we have to use hint

Have you tried using chained mocks?

image

I want you to be using the `assertThat` methods from `assertJ` on the values returned from these methods, if a method returns a flowable you can call `.test()` on it to get back a testSubscriber that has a bunch of assertion methods you can call like `assertValues`

**Comment by akkywadhwa** [comment]
> image
> 
> Sweet sweet coverage.
> 
> We aren't at 100 yet though and that is what I would like to see.
> 
> Also I think you may have overlooked a few of my comments (github can get very messy)
> 
> > I was lazy in my example tests and didn't use chained mocks properly so I gave up on them due to limited time but I found myself on the documentation and we really should use them as they'd make these tests a lot prettier.
> > The trick with chained calls is that for generic return types we have to use hint
> 
> Have you tried using chained mocks?
> 
> image
> 
> I want you to be using the `assertThat` methods from `assertJ` on the values returned from these methods, if a method returns a flowable you can call `.test()` on it to get back a testSubscriber that has a bunch of assertion methods you can call like `assertValues`

I am able to reach 92% of the coverage. 
image

Could you please give some insight on how to test "migrationInsert' properly?
image

Also, I am unable to figure out which branch is remaining in "equal(blah blah"
image

**Comment by macgills** [comment]
@akkywadhwa please push your changes so I can see if there are execution paths missing

**Comment by akkywadhwa** [comment]
> @akkywadhwa please push your changes so I can see if there are execution paths missing

I am still left with 'chained calls'. Anyway, I am pushing.

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 8]
```diff
@@ -51,15 +53,69 @@ internal class NewBookmarksDaoTest {
   @Test
   fun `get bookmarks without current book`() {
     val zimFileReader: ZimFileReader = mockk()
+    val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
+    every { box.query() } returns queryBuilder
+    every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
+    val query: Query<BookmarkEntity> = mockk(relaxed = true)
+    every { queryBuilder.build() } returns query
+    val id = 0L
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:61]: you don't need to declare these vals

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 9]
```diff
@@ -51,15 +53,69 @@ internal class NewBookmarksDaoTest {
...
+    val zimId = "zimId"
+    val zimName = "zimName"
+    val zimFilePath = "zimFilePath"
+    val bookmarkUrl = "bookmarkUrl"
+    val bookmarkTitle = "bookmarkTitle"
+    val favicon: String? = "favicon"
+    val bookmarkEntities: List<BookmarkEntity> = listOf(
+      bookmarkEntity(
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:69]: I wouldn't bother with all this, just write `bookmarkEntity()`

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 10]
```diff
@@ -51,15 +53,69 @@ internal class NewBookmarksDaoTest {
...
+        zimId = zimId,
+        zimName = zimName,
+        zimFilePath = zimFilePath,
+        bookmarkUrl = bookmarkUrl,
+        bookmarkTitle = bookmarkTitle,
+        favicon = favicon
+      )
+    )
+    every { query.find() } returns bookmarkEntities
+    val bookmarkItem: BookmarkItem = mockk()
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:80]: don't mock this, create it. It has a constructor that take a BookmarkEntity

```
  constructor(entity: BookmarkEntity) : this(
    entity.id,
    entity.zimId,
    entity.zimName,
    entity.zimFilePath,
    entity.bookmarkUrl,
    entity.bookmarkTitle,
    entity.favicon
  )
```

**Comment by akkywadhwa** [comment]
> > image
> > Sweet sweet coverage.
> > We aren't at 100 yet though and that is what I would like to see.
> > Also I think you may have overlooked a few of my comments (github can get very messy)
> > > I was lazy in my example tests and didn't use chained mocks properly so I gave up on them due to limited time but I found myself on the documentation and we really should use them as they'd make these tests a lot prettier.
> > > The trick with chained calls is that for generic return types we have to use hint
> > 
> > 
> > Have you tried using chained mocks?
> > image
> > I want you to be using the `assertThat` methods from `assertJ` on the values returned from these methods, if a method returns a flowable you can call `.test()` on it to get back a testSubscriber that has a bunch of assertion methods you can call like `assertValues`
> 
> I am able to reach 92% of the coverage.
> 
> image
> 
> Could you please give some insight on how to test "migrationInsert' properly? Will this require mocking the functionality of `bookDao.getFavIconAndZimFile()`?
> 
> image
> 
> Also, I am facing difficulty to figure out which branch is remaining in "equal(blah blah)"
> 
> image
> 
> Tag: @macgills

@macgills what about the remaining branches of code shown in the screenshot?

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 11]
```diff
@@ -51,15 +53,69 @@ internal class NewBookmarksDaoTest {
...
+          bookmarkUrl = bookmarkUrl,
+          bookmarkTitle = bookmarkTitle,
+          favicon = favicon
+        )
+      )
+    )
+  }
+
+  @Test
+  fun `get bookmarks without current book when zimFileReader is null`() {
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:109]: you only need 1 test without current book to check that execution path, this test should be "with current book"

to test this
```
 if (fromCurrentBook) {
      equal(BookmarkEntity_.zimName, zimFileReader?.name ?: "")
    }
```
we want 1 test `fromCurrentBook=true zimFileReader=null` and another test `fromCurrentBook=true zimFileReader=mockThatWillReturnASpecificName()`

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 12]
```diff
@@ -51,15 +53,69 @@ internal class NewBookmarksDaoTest {
...
     val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
     every { box.query() } returns queryBuilder
     every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
     val query: Query<BookmarkEntity> = mockk(relaxed = true)
     every { queryBuilder.build() } returns query
     val bookmarkEntities: List<BookmarkEntity> = mockk(relaxed = true)
     every { query.find() } returns bookmarkEntities
-    newBookmarksDao.getBookmarks(false, zimFileReader)
-    verify { queryBuilder.build() }
+    newBookmarksDao.getBookmarks(false, null)
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:117]: all test methods should have an assertThat if the method returns a value

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 13]
```diff
@@ -152,6 +214,31 @@ internal class NewBookmarksDaoTest {
...
+    every {
+      queryBuilder.equal(
+        BookmarkEntity_.zimId,
+        ""
+      )
+    } returns queryBuilder
+    every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
+    val query: Query<BookmarkEntity> = mockk(relaxed = true)
+    every { queryBuilder.build() } returns query
+    newBookmarksDao.bookmarkUrlsForCurrentBook(null).test()
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:236]:
```
    .property(BookmarkEntity_.bookmarkUrl)
    .findStrings()
    .toList()
    .distinct()
```
none of this is tested

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 14]
```diff
@@ -82,47 +138,53 @@ internal class NewBookmarksDaoTest {
   }
 
   @Test
-  fun `get bookmarks from current book when zimFileReader is null`() {
-    val zimFileReader: ZimFileReader? = null
+  fun `get current zim bookmarks url`() {
+    val zimFileReader: ZimFileReader = mockk()
     val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
     every { box.query() } returns queryBuilder
+    every { zimFileReader.name } returns ""
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:145]: neither tests mention id and the name value you return here should be different from what the value is normally
```
    equal(BookmarkEntity_.zimId, zimFileReader?.id ?: "")
      .or()
      .equal(BookmarkEntity_.zimName, zimFileReader?.name ?: "")
```

I also wouldn't reuse the mock 

```
queryBuilder.equal(
        BookmarkEntity_.zimName,
        zimFileReader?.name ?: ""
        zimFileReader.name
      )
```

for this call, I would extract a variable

```
val expectedName = "zimName"
every {zimFileReader.name} returns expectedName
   every {
      queryBuilder.equal(
        BookmarkEntity_.zimName,
        expectedName
      )
    } returns queryBuilder
```

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 15]
```diff
@@ -174,9 +261,38 @@ internal class NewBookmarksDaoTest {
...
+            zimId = zimId,
+            zimName = zimName,
+            zimFilePath = zimFilePath,
+            bookmarkUrl = bookmarkUrl,
+            bookmarkTitle = bookmarkTitle,
+            favicon = favicon
+          )
+        )
+      )
+    }
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:293]: for the test below you need to expect data from
`bookDao.getFavIconAndZimFile(bookMark)`

don't mock data
```
val bookmarks: MutableList<Bookmark> = mockk(relaxed = true)
```

create it

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+    verify { bookmarkEntities.map(::BookmarkItem) }
+  }
+
+  @Test
+  fun `get bookmarks from current book`() {
+    val zimFileReader: ZimFileReader = mockk()
+    val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
+    every { box.query() } returns queryBuilder
+    every { zimFileReader.name } returns ""
+    every {
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:127]: There is a lot of formatting like this that could fit on one line

```  
  every { queryBuilder.equal(BookmarkEntity_.zimName, zimFileReader.name) } returns queryBuilder
```

**Comment by macgills** [comment]
Don't bother with using `hint`, it is only useful if the return type is generic, not if it is a type parameter being erased

**Comment by akkywadhwa** [comment]
@macgills I have made the changes as suggested. The coverage has reduced to 80% because `migrationInsert` is being considered as tested now. Could you please advice how to take this further?

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+            bookmarkTitle = bookmarkTitle,
+            favicon = favicon
+          )
+        )
+      )
+    }
+  }
+
+  @Test
+  fun `migrate insert`() {
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:254]: how is any of the following code under test from this method?

```
  fun migrationInsert(
    bookmarks: MutableList<Bookmark>,
    bookDao: NewBookDao
  ) {
    box.put(bookmarks.zip(bookmarks.map(bookDao::getFavIconAndZimFile)).map(::BookmarkEntity))
  }
```

You need to be verifying `box.put(expectedEntities)`

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+  }
+
+  @Test
+  fun `get bookmarks from current book`() {
+    val zimFileReader: ZimFileReader = mockk()
+    val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
+    every { box.query() } returns queryBuilder
+    val zimName = "zimName"
+    every { zimFileReader?.name } returns zimName
+    every { queryBuilder.equal(BookmarkEntity_.zimName, zimFileReader?.name) } returns queryBuilder
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:94]: if you were to verify this step  becausewas called then this test would fail because

```
newBookmarksDao.getBookmarks(false, zimFileReader)
```

you passed in false

Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:94]: this is a relic of copy pasting code? zimFileReader cannot be null here so `?` is pointless and I had a comment elsewhere that I would prefer mocks to not be reused like this, the variable `zimName` should be used instead

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+    assertThat(newBookmarksDao.getBookmarks(false, zimFileReader))
+      .isEqualTo(listOf(bookmarkItem))
+  }
+
+  @Test
+  fun `getCurrentZimBookmarksUrl returns list of bookmarkUrl`() {
+    val zimFileReader: ZimFileReader = mockk()
+    val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
+    every { box.query() } returns queryBuilder
+    val zimId = ""
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:110]: these values are indistinct from whats get returned if ZimFileReader is null making the test less legible

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+    val propertyQuery: PropertyQuery = mockk()
+    every { query.property(BookmarkEntity_.bookmarkUrl) } returns propertyQuery
+    val bookmarkUrlArray = arrayOf<String>(bookmarkUrl, bookmarkUrl)
+    every { propertyQuery.findStrings() } returns bookmarkUrlArray
+    assertThat(newBookmarksDao.getCurrentZimBookmarksUrl(zimFileReader))
+      .isEqualTo(listOf(bookmarkUrl))
+  }
+
+  @Test
+  fun `getCurrentZimBookmarksUrl returns empty list with empty zimName and zimId`() {
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:154]: `searches for empty name & id with a null zimFileReader` may be more accurate here

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt (hunk 13 above)
Comment (macgills) [reply core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:193]: tch, this was  stupid comment, looking at the wrong method

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+    every { queryBuilder.build() } returns query
+    assertThat(newBookmarksDao.getCurrentZimBookmarksUrl(null)).isEmpty()
+  }
+
+  @Test
+  fun `bookmark urls for current workbook`() {
+    val zimFileReader: ZimFileReader = mockk()
+    val queryBuilder: QueryBuilder<BookmarkEntity> = mockk()
+    every { box.query() } returns queryBuilder
+    every { zimFileReader.name } returns ""
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:171]: indistinct from default values

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+    every { queryBuilder.equal(BookmarkEntity_.zimName, zimFileReader.name) } returns queryBuilder
+    every { queryBuilder.or() } returns queryBuilder
+    every { zimFileReader.id } returns ""
+    every { queryBuilder.equal(BookmarkEntity_.zimId, zimFileReader.id) } returns queryBuilder
+    every { queryBuilder.order(BookmarkEntity_.bookmarkTitle) } returns queryBuilder
+    val query: Query<BookmarkEntity> = mockk(relaxed = true)
+    every { queryBuilder.build() } returns query
+    newBookmarksDao.bookmarkUrlsForCurrentBook(zimFileReader).test()
+      .assertEmpty()
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:180]: I want to see more than `assertEmpty` you can look at RecentSearchDaoTest

```
      val queryBuilder = mockk<QueryBuilder<RecentSearchEntity>>()
      every { box.query() } returns queryBuilder
      every { queryBuilder.equal(RecentSearchEntity_.zimId, zimId) } returns queryBuilder
      every { queryBuilder.orderDesc(RecentSearchEntity_.id) } returns queryBuilder
      val query = mockk<Query<RecentSearchEntity>>()
      every { queryBuilder.build() } returns query
      mockkStatic(RxQuery::class)
      every { RxQuery.observable(query) } returns Observable.just(queryResult)
    }
```

for an example of how you deal with `box.asFlowable`

File: core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt (hunk 15 above)
Comment (macgills) [reply core/src/test/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDaoTest.kt:250]: @akkywadhwa I was completely wrong here, forgot Bookmark is a TableModel ugh. mock is totally acceptable but you would want `listOf(mockk())`

**Comment by macgills** [comment]
@akkywadhwa do you have the availability to finish this PR? If not I will close it next time it is marked stale, in 7 days.

**Comment by macgills** [comment]
Closing due to inactivity.
